pkg/volume/provision.go

```diff
@@ -447,8 +475,24 @@ func (p *nfsProvisioner) getServer() (string, error) {
 		return "", fmt.Errorf("service %s=%s is valid but it doesn't have a cluster IP", p.serviceEnv, serviceName)
 	}
 
-	glog.Infof("using service %s=%s cluster IP %s as NFS server IP", p.serviceEnv, serviceName, service.Spec.ClusterIP)
-	return service.Spec.ClusterIP, nil
+	// do the nodes support "cluster.local" internal DNS names?
+	// TODO maybe we can auto-detect this, or make this user-configurable?
+	nodes_support_cluster_local := false
+
+	// in a perfect world we would like to hand out a "service.namespace.svc.cluster.local" address,
+	// because then the NFS mount point will also work in case of a disaster (service is recreated
+	// with a new IP). However, because the "mount -t nfs" command is given on the nodes and not 
+	// inside Kubernetes, such DNS names do not exist on normally configured clusters.
+	if nodes_support_cluster_local {
+		// we go for the namespace name and service name, and rely on the node to find the IP address
+		glog.Infof("using service %s=%s hostname %s.%s as NFS server IP", p.serviceEnv, serviceName, serviceName, namespace)
+		return serviceName + "." + namespace + ".svc.cluster.local", nil
```

Comment:
Same as above. you can't grantee that cluster is using this FQDN, however the shorten name should also work
```suggestion
		return serviceName + "." + namespace + ".svc", nil
```

Reply:
I am not entirely sure that 

> The main idea, that service name is providing you stable cluster IP which is never changes until you delete the service itself.
> 
> nfs-server-provisioner is reading this clusterIP from the service, because it is used to generate PersistentVolumes, which are also immutable.
> 
> If you want to add an opportunity using DNS-hostname instead, I would simple specify `--server-hostname` option, this hostname may also be used for generating `PersistentVolumes` bypassing the need of having separate service in Kubernetes.

Yes, that is exactly what I ended up doing after a few iterations of the code. So this pull request now comes down to "make sure that --server-hostname also works when running *in* the cluster" (it currently only seems to work when running *outside* the cluster).

The other changes are mainly cosmetical, and can be used as inspiration to further clean up the code (because there are some things that I find odd, especially the code that uses the unreliable node-name or pod-ip as nfs address)